### DISCUSSION TITLE
خطا در استفاده از useDistributedCache

### DISCUSSION THREAD

**Discussion by SaeedNoori77** [Sina-Soltani/Parbad#361] [Q&A]
سلام وقت بخیر
توی داکیومنت های پکیج گفتین توی محیط پروداکشن به جای useMemoryCache میتونین از useDistributedCache استفاده کنین
من ردیس رو به پروژم اضافه کردم و از useDistributedCache استفاده کردم ولی خطا میده
اگه ممکنه یکم بیشتر توضیح بدین درمورد استفاده از useDistributedCache

**Comment by Sina-Soltani** [comment]
درود
لطفا متن خطا و توضیحات بیشتر در اون رابطه رو قرار بدید تا بتونم کمکی بکنم.
بدون هیچ جزئیاتی که کمکی از دست من بر نمیاد.

**Comment by SaeedNoori77** [comment]
Autofac.Core.DependencyResolutionException: An exception was thrown while activating Parbad.Internal.DefaultOnlinePayment -> Parbad.Storage.StorageManager -> Parbad.Storage.Cache.DistributedCache.DistributedCacheStorage.\r\n ---> Autofac.Core.DependencyResolutionException: An exception was thrown while invoking the constructor 'Void .ctor(Microsoft.Extensions.Caching.Distributed.IDistributedCache, Microsoft.Extensions.Options.IOptions`1[Parbad.Storage.Cache.DistributedCache.DistributedCacheStorageOptions])' on type 'DistributedCacheStorage'.\r\n ---> Newtonsoft.Json.JsonSerializationException: Could not create an instance of type Parbad.Storage.Cache.Abstractions.ICacheStorageCollection. Type is an interface or abstract class and cannot be instantiated. Path 'Payments', line 1, position 12.\r\n   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewObject(JsonReader reader, JsonObjectContract objectContract, JsonProperty containerMember, JsonProperty containerProperty, String id, Boolean& createdFromNonDefaultCreator)\r\n   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)\r\n   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)\r\n   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)\r\n   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)\r\n   at Newtonsoft.Json.JsonSerializer.Deserialize(JsonReader reader, Type objectType)\r\n   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)\r\n   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)\r\n   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value)\r\n   at Parbad.Storage.Cache.Internal.ObjectSerializer.DeserializeObject[T](Byte[] buffer)\r\n   at Parbad.Storage.Cache.DistributedCache.DistributedCacheStorage.BuildCollection()\r\n   at Parbad.Storage.Cache.DistributedCache.DistributedCacheStorage..ctor(IDistributedCache distributedCache, IOptions`1 options)\r\n   at lambda_method10(Closure , Object[] )\r\n   at Autofac.Core.Activators.Reflection.BoundConstructor.Instantiate()\r\n   --- End of inner exception stack trace ---\r\n   at Autofac.Core.Activators.Reflection.BoundConstructor.Instantiate()\r\n   at Autofac.Core.Activators.Reflection.ReflectionActivator.<>c__DisplayClass12_0.<UseSingleConstructorActivation>b__0(ResolveRequestContext ctxt, Action`1 next)\r\n   at Autofac.Core.Resolving.Middleware.DelegateMiddleware.Execute(ResolveRequestContext context, Action`1 next)\r\n   at Autofac.Core.Resolving.Pipeline.ResolvePipelineBuilder.<>c__DisplayClass14_0.<BuildPipeline>b__1(ResolveRequestContext ctxt)\r\n   at Autofac.Core.Resolving.Middleware.DisposalTrackingMiddleware.Execute(ResolveRequestContext context, Action`1 next)\r\n   at Autofac.Core.Resolving.Pipeline.ResolvePipelineBuilder.<>c__DisplayClass14_0.<BuildPipeline>b__1(ResolveRequestContext ctxt)\r\n   at Autofac.Core.Resolving.Middleware.ActivatorErrorHandlingMiddleware.Execute(ResolveRequestContext context, Action`1 next)\r\n   --- End of inner exception stack trace ---\r\n   at Autofac.Core.Resolving.Middleware.ActivatorErrorHandlingMiddleware.Execute(ResolveRequestContext context, Action`1 next)\r\n   at Autofac.Core.Resolving.Pipeline.ResolvePipelineBuilder.<>c__DisplayClass14_0.<BuildPipeline>b__1(ResolveRequestContext ctxt)\r\n   at Autofac.Core.Pipeline.ResolvePipeline.Invoke(ResolveRequestContext ctxt)\r\n   at Autofac.Core.Resolving.Middleware.RegistrationPipelineInvokeMiddleware.Execute(ResolveRequestContext context, Action`1 next)\r\n   at Autofac.Core.Resolving.Pipeline.ResolvePipelineBuilder.<>c__DisplayClass14_0.<BuildPipeline>b__1(ResolveRequestContext ctxt)\r\n   at Autofac.Core.Resolving.Middleware.SharingMiddleware.Execute(ResolveRequestContext context, Action`1 next)\r\n   at Autofac.Core.Resolving.Pipeline.ResolvePipelineBuilder.<>c__DisplayClass14_0.<BuildPipeline>b__1(ResolveRequestContext ctxt)\r\n   at Autofac.Core.Resolving.Middleware.ScopeSelectionMiddleware.Execute(ResolveRequestContext context, Action`1 next)\r\n   at Autofac.Core.Resolving.Pipeline.ResolvePipelineBuilder.<>c__DisplayClass14_0.<BuildPipeline>b__1(ResolveRequestContext ctxt)\r\n   at Autofac.Core.Resolving.Middleware.CircularDependencyDetectorMiddleware.Execute(ResolveRequestContext context, Action`1 next)\r\n   at Autofac.Core.Resolving.Pipeline.ResolvePipelineBuilder.<>c__DisplayClass14_0.<BuildPipeline>b__1(ResolveRequestContext ctxt)\r\n   at Autofac.Core.Pipeline.ResolvePipeline.Invoke(ResolveRequestContext ctxt)\r\n   at Autofac.Core.Resolving.ResolveOperation.GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, ResolveRequest request)\r\n   at Autofac.Core.Resolving.ResolveOperation.ExecuteOperation(ResolveRequest request)\r\n   at Autofac.Core.Resolving.ResolveOperation.Execute(ResolveRequest request)\r\n   at Autofac.Core.Lifetime.LifetimeScope.ResolveComponent(ResolveRequest request)\r\n   at Autofac.ResolutionExtensions.TryResolveService(IComponentContext context, Service service, IEnumerable`1 parameters, Object& instance)\r\n   at Autofac.ResolutionExtensions.ResolveOptionalService(IComponentContext context, Service service, IEnumerable`1 parameters)\r\n   at Autofac.ResolutionExtensions.ResolveOptional(IComponentContext context, Type serviceType, IEnumerable`1 parameters)\r\n   at Autofac.ResolutionExtensions.ResolveOptional(IComponentContext context, Type serviceType)\r\n   at Autofac.Extensions.DependencyInjection.AutofacServiceProvider.GetService(Type serviceType)\r\n   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetService(IServiceProvider sp, Type type, Type requiredBy, Boolean isDefaultParameterRequired)\r\n   at lambda_method431(Closure , IServiceProvider , Object[] )\r\n   at Microsoft.AspNetCore.Mvc.Controllers.ControllerActivatorProvider.<>c__DisplayClass7_0.<CreateActivator>b__0(ControllerContext controllerContext)\r\n   at Microsoft.AspNetCore.Mvc.Controllers.ControllerFactoryProvider.<>c__DisplayClass6_0.<CreateControllerFactory>g__CreateController|0(ControllerContext controllerContext)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()\r\n--- End of stack trace from previous location ---\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|26_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)

**Reply by SaeedNoori77** [reply]
به این شکل ردیس رو اضافه کردم

var redisConnectionString = Configuration.GetConnectionString("redis");
services.AddStackExchangeRedisCache(o => { o.Configuration = redisConnectionString; });

طبق داکیومنت های پکیج هم این کد رو نوشتم

services.AddParbad()
                .ConfigureGateways(gateways =>
                {
                    gateways.AddMelli()
                        .WithOptions(opt =>
                        {
                            opt.ApiRequestUrl =
                                "https://sandbox.banktest.ir/melli/sadad.shaparak.ir/VPG/api/v0/Request/PaymentRequest";
                            opt.ApiVerificationUrl =
                                "https://sandbox.banktest.ir/melli/sadad.shaparak.ir/VPG/api/v0/Advice/Verify";
                            opt.PaymentPageUrl = "https://sandbox.banktest.ir/melli/sadad.shaparak.ir/VPG/Purchase";
                        })
                        .WithAccounts(accounts =>
                        {
                            accounts.AddInMemory(account =>
                            {
                                account.MerchantId = "MerchantId";
                                account.TerminalId = "TerminalId";
                                account.TerminalKey = "TerminalKey";
                            });
                        });
                    gateways.AddParbadVirtual()
                        .WithOptions(options => options.GatewayPath = "/MyVirtualGateway");
                })
                .ConfigureHttpContext(builder => builder.UseDefaultAspNetCore())
                .ConfigureStorage(builder => builder.UseDistributedCache());

**Comment by Sina-Soltani** [comment]
ظاهرا مشکلی که وجود داره اینه که شما اولین قدم تنظیم و در واقع اولین قسمت شروع تنظیمات که در مستندات گفته شده رو انجام ندادید.
شما به جای استفاده از تزریق وابستگی داخلی دات نت، از آتو فک که یک کتابخانه جانبی هست استفاده کردید در پروژه خودتون.
قسمت اول تنظیمات در مستندات برای شما توضیح داده که اگر قصد استفاده از تزریق وابستگی جانبی ای رو دارید، باید اون رو تطابق بدید و معرفی کنید.
علت اینکه شما چنین خطایی دریافت میکنید اینه که تنظیمات رو دارید به پَرباد معرفی میکنید، ولی آتو فک هیچ خبری ازشون نداره.
لطفا مستندات رو نگاه کنید.
[Using a Dependency Injection library (such as Autofac)](#using-a-dependency-injection-library-such-as-autofac)

**Reply by SaeedNoori77** [reply]
با این هم همون نتیجه رو میگیرم

var parbadBuilder = ParbadBuilder.CreateDefaultBuilder()
                .ConfigureGateways(gateways =>
                    {
                        gateways.AddMelli()
                            .WithOptions(opt =>
                            {
                                opt.ApiRequestUrl =
                                    "https://sandbox.banktest.ir/melli/sadad.shaparak.ir/VPG/api/v0/Request/PaymentRequest";
                                opt.ApiVerificationUrl =
                                    "https://sandbox.banktest.ir/melli/sadad.shaparak.ir/VPG/api/v0/Advice/Verify";
                                opt.PaymentPageUrl = "https://sandbox.banktest.ir/melli/sadad.shaparak.ir/VPG/Purchase";
                            })
                            .WithAccounts(accounts =>
                            {
                                accounts.AddInMemory(account =>
                                {
                                    account.MerchantId = "MerchantId";
                                    account.TerminalId = "TerminalId";
                                    account.TerminalKey = "TerminalKey";
                                });
                            });
                        gateways.AddParbadVirtual()
                            .WithOptions(options => options.GatewayPath = "/MyVirtualGateway");
                    })
                    .ConfigureHttpContext(builder => builder.UseDefaultAspNetCore())
                    .ConfigureStorage(builder => builder.UseDistributedCache());
            builder.Populate(parbadBuilder.Services);

**Reply by Sina-Soltani** [reply]
ظاهرا مشکل به خاطر استفاده از یک اینترفیس بوده در deserialize کردن داده ها که در واقع باید کلاس استفاده میشده به جای اینترفیس.
فردا یک نسخه برای رفع این مشکل منتشر میکنم.

**Reply by SaeedNoori77** [reply]
ممنون از شما

**Comment by Sina-Soltani** [comment]
لطفا نسخه جدید رو تست کنید و نتیجه رو اعلام کنید

https://www.nuget.org/packages/Parbad.Storage.Cache/1.3.1-beta1

**Reply by SaeedNoori77** [reply]
مشکل موجود رفع شد. و جالبه که بدون استفاده از Using a Dependency Injection library (such as Autofac) کار میکنه.
یک مشکل دیگه مربوط به آدرس درگاه ملی رو مطرح کرده بودم. برای رفع این مورد باید منتظر نسخه غیر بتا باشم؟

**Comment by Sina-Soltani** [comment]
نسخه بتا ای که برای تست بهتون گفتم رو تبدیل میکنم به نسخه عادی و منتشر میکنم.
برای تصحیح آدرس درگاه ملی یک نسخه بتا منتشر میکنم و مانند همین روال بهتون اطلاع میدم که تست کنید و نتیجه رو اعلام کنید.

**Reply by SaeedNoori77** [reply]
تشکر

**Reply by Sina-Soltani** [reply]
https://www.nuget.org/packages/Parbad.Storage.Cache/1.3.1
